File: src/lira/linguistics/role/sentence_reader.py
"""SentenceReader: spec 14.3's actual `Sentence.read()` implementation.
Splits off trailing punctuation, reads the remainder as one
ClauseType.INDEPENDENT clause (ClauseReader -- Phase 1 supports exactly
one clause per sentence, see clause_reader.py's own module docstring),
and checks the result against
GrammarConfigurator.sentence_templates[SentenceType.DECLARATIVE], the
only sentence template populated in this phase (sentence_type.py).
Accepts either raw text (tokenised via TokenResolver.resolve_sentence,
as exactly one sentence -- splitting a longer string into several
sentences is LinguisticController.read_text's job, not this method's)
or an already-resolved TokenReading sequence (spec 14.3's "text or
pre-resolved tokens")."""

from typing import List, Optional, Sequence, Union

from ..data.linguistic_unit_kind import LinguisticUnitKind
from ..data.reading_error import ReadingError, ReadingErrorKind
from ..data.sentence import Sentence
from ..data.sentence_type import SentenceType
from ..data.token_reading import TokenReading
from ..data.validation_outcome import ValidationOutcome
from .clause_reader import ClauseReader
from .grammar_configurator import GrammarConfigurator
from .reading_scorer import ScoringFactors
from .sequence_engine import SequenceEngine
from .token_resolver import TokenResolver


class SentenceReader:
    def __init__(
        self, clause_reader: ClauseReader, token_resolver: TokenResolver, engine: SequenceEngine, grammar: GrammarConfigurator,
    ):
        self.clause_reader = clause_reader
        self.token_resolver = token_resolver
        self.engine = engine
        self.grammar = grammar

    def read(
        self,
        text_or_tokens: Union[str, Sequence[TokenReading]],
        *,
        grammar: Optional[GrammarConfigurator] = None,
        sequence_number: int = 0,
        trace: Optional[List[dict]] = None,
    ) -> Sentence:
        """`trace`, when a list is passed, is threaded straight through
        to ClauseReader.read -- see PhraseReader.read's own docstring
        for what gets recorded. Purely additive/observational."""
        active_grammar = grammar or self.grammar
        tokens = (
            self.token_resolver.resolve_sentence(text_or_tokens)
            if isinstance(text_or_tokens, str)
            else tuple(text_or_tokens)
        )

        if not tokens:
            return self._empty_sentence(sequence_number)

        punctuation_token = tokens[-1]
        has_terminal_punctuation = punctuation_token.is_punctuation
        clause_end = len(tokens) - 1 if has_terminal_punctuation else len(tokens)

        clause = self.clause_reader.read(tokens, start_index=0, end_index=clause_end, grammar=active_grammar, trace=trace)

        punctuation_word = None
        if has_terminal_punctuation:
            selected = punctuation_token.candidates[0] if punctuation_token.candidates else None
            punctuation_word = self.clause_reader.phrase_reader.graph_processor.materialise_token(
                punctuation_token, punctuation_token.token_index, selected_candidate=selected,
            )

        sentence_template = active_grammar.sentence_templates.get(SentenceType.DECLARATIVE)
        errors: List[ReadingError] = list(clause.errors)
        outcome = clause.validation
        sentence_type = None

        if sentence_template is None:
            outcome = ValidationOutcome.INVALID
            errors.append(ReadingError(
                kind=ReadingErrorKind.NO_VALID_SENTENCE_INTERPRETATION, level=LinguisticUnitKind.Sentence,
                message="No sentence template configured for SentenceType.DECLARATIVE",
            ))
        else:
            sentence_type = SentenceType.DECLARATIVE
            if punctuation_word is not None and punctuation_word.text not in sentence_template.terminal_punctuation:
                outcome = min(outcome, ValidationOutcome.INVALID, key=lambda o: o.value)
                errors.append(ReadingError(
                    kind=ReadingErrorKind.INVALID_PUNCTUATION_SEQUENCE, level=LinguisticUnitKind.Sentence,
                    message=f'"{punctuation_word.text}" is not valid terminal punctuation for a declarative sentence',
                    token_index=punctuation_token.token_index, token_text=punctuation_token.text,
                ))

        sentence_words = list(clause.tokens) + ([punctuation_word] if punctuation_word is not None else [])
        factors = ScoringFactors(
            validation=outcome,
            unresolved_token_count=sum(1 for token in tokens if not token.is_known),
            undischarged_obligation_count=sum(len(phrase.open_obligations) for phrase in clause.phrases),
            finite_verb_phrase_count=1 if clause.predicate is not None else 0,
            phrase_count=len(clause.phrases),
            lexical_evidence_sum=clause.confidence,
        )
        confidence = self.engine.scorer.confidence(factors)

        sentence = Sentence(
            text=" ".join(word.text for word in sentence_words),
            clauses=[clause],
            requires_punctuation=bool(sentence_template.terminal_punctuation) if sentence_template else None,
            tokens=sentence_words,
            sentence_type=sentence_type,
            selected_parts_of_speech=tuple(word.part_of_speech for word in sentence_words),
            punctuation=punctuation_word,
            validation=outcome,
            confidence=confidence,
            errors=tuple(errors),
        )
        sentence.system_property = self.clause_reader.phrase_reader.graph_processor.create_property_wrapper(
            sentence, LinguisticUnitKind.Sentence, sequence_number, "SentenceReader_ReadLayer",
        )
        return sentence

    def _empty_sentence(self, sequence_number: int = 0) -> Sentence:
        error = ReadingError(
            kind=ReadingErrorKind.NO_VALID_SENTENCE_INTERPRETATION, level=LinguisticUnitKind.Sentence,
            message="Empty token sequence",
        )
        sentence = Sentence(text="", clauses=[], validation=ValidationOutcome.INVALID, confidence=0.0, errors=(error,))
        sentence.system_property = self.clause_reader.phrase_reader.graph_processor.create_property_wrapper(
            sentence, LinguisticUnitKind.Sentence, sequence_number, "SentenceReader_ReadLayer",
        )
        return sentence

File: prototype/src/lira/linguistics/role/sentence_reader.ts
import { LinguisticUnitKind } from "../data/linguistic_unit_kind";
import { createReadingError, ReadingErrorKind, type ReadingError } from "../data/reading_error";
import { createSentence, type Sentence } from "../data/sentence";
import { SentenceType } from "../data/sentence_type";
import { isKnown, isPunctuation, type TokenReading } from "../data/token_reading";
import { ValidationOutcome } from "../data/validation_outcome";
import type { ClauseReader } from "./clause_reader";
import type { GrammarConfigurator, SentenceTemplate } from "./grammar_configurator";
import { createScoringFactors } from "./reading_scorer";
import type { SequenceEngine } from "./sequence_engine";
import type { TokenResolver } from "./token_resolver";

/** SentenceReader: spec 14.3's actual readSentence() implementation.
 * Splits off trailing punctuation, reads the remainder as one
 * ClauseType.INDEPENDENT clause (ClauseReader -- Phase 1 supports
 * exactly one clause per sentence), and checks the result against
 * whichever of GrammarConfigurator.sentenceTemplates actually matches
 * the sentence's own terminal punctuation mark (DECLARATIVE "."/
 * INTERROGATIVE "?"/EXCLAMATORY "!" as of this phase -- see
 * selectSentenceTemplate below and grammar_configurator.ts's own
 * buildSentenceTemplates docstring on why all three share DECLARATIVE's
 * exact clause shape rather than each enforcing distinct word-order
 * grammar). Accepts either raw text (tokenised via
 * TokenResolver.resolveSentence, as exactly one sentence) or an
 * already-resolved TokenReading sequence.
 *
 * Ported from linguistics/role/sentence_reader.py, extended
 * (prototype only) with the INTERROGATIVE/EXCLAMATORY template
 * matching above -- Python's own sentence_reader.py still hardcodes
 * SentenceType.DECLARATIVE (see this session's standing TypeScript-only
 * scope). */

export interface SentenceReadOptions {
  grammar?: GrammarConfigurator;
  sequenceNumber?: number;
  trace?: unknown[];
}

export class SentenceReader {
  constructor(
    public readonly clauseReader: ClauseReader,
    private readonly tokenResolver: TokenResolver,
    private readonly engine: SequenceEngine,
    private readonly grammar: GrammarConfigurator,
  ) {}

  /** `trace`, when an array is passed, is threaded straight through to
   * ClauseReader.read. Purely additive/observational. */
  read(textOrTokens: string | readonly TokenReading[], options: SentenceReadOptions = {}): Sentence {
    const activeGrammar = options.grammar ?? this.grammar;
    const sequenceNumber = options.sequenceNumber ?? 0;
    const tokens = typeof textOrTokens === "string"
      ? this.tokenResolver.resolveSentence(textOrTokens)
      : textOrTokens;

    if (tokens.length === 0) return this.emptySentence(sequenceNumber);

    const punctuationToken = tokens[tokens.length - 1];
    const hasTerminalPunctuation = isPunctuation(punctuationToken);
    const clauseEnd = hasTerminalPunctuation ? tokens.length - 1 : tokens.length;

    const clause = this.clauseReader.read(tokens, { startIndex: 0, endIndex: clauseEnd, grammar: activeGrammar, trace: options.trace });

    let punctuationWord;
    if (hasTerminalPunctuation) {
      const selected = punctuationToken.candidates.length > 0 ? punctuationToken.candidates[0] : undefined;
      punctuationWord = this.clauseReader.phraseReader.graphProcessor.materialiseToken(
        punctuationToken, punctuationToken.tokenIndex, selected,
      );
    }

    const sentenceTemplate = this.selectSentenceTemplate(activeGrammar, punctuationWord?.text);
    const errors: ReadingError[] = [...clause.errors];
    let outcome = clause.validation;
    let sentenceType: SentenceType | undefined;

    if (sentenceTemplate === undefined) {
      outcome = ValidationOutcome.INVALID;
      errors.push(createReadingError({
        kind: ReadingErrorKind.NO_VALID_SENTENCE_INTERPRETATION, level: LinguisticUnitKind.Sentence,
        message: "No sentence template configured for SentenceType.DECLARATIVE",
      }));
    } else {
      sentenceType = sentenceTemplate.sentenceType;
      if (punctuationWord !== undefined && !sentenceTemplate.terminalPunctuation.has(punctuationWord.text)) {
        outcome = outcome < ValidationOutcome.INVALID ? outcome : ValidationOutcome.INVALID;
        errors.push(createReadingError({
          kind: ReadingErrorKind.INVALID_PUNCTUATION_SEQUENCE, level: LinguisticUnitKind.Sentence,
          message: `"${punctuationWord.text}" is not valid terminal punctuation for any configured sentence type`,
          tokenIndex: punctuationToken.tokenIndex, tokenText: punctuationToken.text,
        }));
      }
    }

    const sentenceWords = punctuationWord !== undefined ? [...clause.tokens, punctuationWord] : [...clause.tokens];
    const factors = createScoringFactors({
      validation: outcome,
      unresolvedTokenCount: tokens.filter((token) => !isKnown(token)).length,
      undischargedObligationCount: clause.phrases.reduce((sum, phrase) => sum + phrase.openObligations.length, 0),
      finiteVerbPhraseCount: clause.predicate !== undefined ? 1 : 0,
      phraseCount: clause.phrases.length,
      lexicalEvidenceSum: clause.confidence,
    });
    const confidence = this.engine.scorer.confidence(factors);

    const sentence = createSentence({
      text: sentenceWords.map((word) => word.text).join(" "),
      clauses: [clause],
      requiresPunctuation: sentenceTemplate !== undefined ? sentenceTemplate.terminalPunctuation.size > 0 : undefined,
      tokens: sentenceWords,
      sentenceType,
      selectedPartsOfSpeech: sentenceWords.map((word) => word.partOfSpeech),
      punctuation: punctuationWord,
      validation: outcome,
      confidence,
      errors,
    });
    sentence.systemProperty = this.clauseReader.phraseReader.graphProcessor.createPropertyWrapper(
      sentence, LinguisticUnitKind.Sentence, sequenceNumber, "SentenceReader_ReadLayer",
    );
    return sentence;
  }

  /** Picks whichever configured SentenceTemplate's own terminalPunctuation
   * actually contains this sentence's terminal punctuation mark ("."/
   * "?"/"!" as of this phase), tried in Map insertion order (DECLARATIVE
   * first, then INTERROGATIVE, then EXCLAMATORY -- grammar_configurator.ts's
   * own buildSentenceTemplates). A sentence with no terminal punctuation
   * token at all (`punctuationText` undefined) defaults to DECLARATIVE --
   * the same permissive behaviour this reader always had; a missing full
   * stop was never itself grounds to reject an otherwise-valid clause. A
   * terminal mark that matches no configured template (e.g. ";") also
   * falls back to DECLARATIVE, purely so the caller's own
   * INVALID_PUNCTUATION_SEQUENCE check has a real template to compare
   * against and report the mismatch -- it is not claiming the sentence
   * actually is declarative. */
  private selectSentenceTemplate(grammar: GrammarConfigurator, punctuationText: string | undefined): SentenceTemplate | undefined {
    if (punctuationText !== undefined) {
      for (const template of grammar.sentenceTemplates.values()) {
        if (template.terminalPunctuation.has(punctuationText)) return template;
      }
    }
    return grammar.sentenceTemplates.get(SentenceType.DECLARATIVE);
  }

  private emptySentence(sequenceNumber = 0): Sentence {
    const error = createReadingError({
      kind: ReadingErrorKind.NO_VALID_SENTENCE_INTERPRETATION, level: LinguisticUnitKind.Sentence,
      message: "Empty token sequence",
    });
    const sentence = createSentence({ text: "", clauses: [], validation: ValidationOutcome.INVALID, confidence: 0.0, errors: [error] });
    sentence.systemProperty = this.clauseReader.phraseReader.graphProcessor.createPropertyWrapper(
      sentence, LinguisticUnitKind.Sentence, sequenceNumber, "SentenceReader_ReadLayer",
    );
    return sentence;
  }
}
